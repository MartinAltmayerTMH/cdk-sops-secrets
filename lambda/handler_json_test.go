package main

import (
	"context"
	"encoding/json"
	"os"
	"testing"
	"time"

	"github.com/aws/aws-lambda-go/cfn"
	"github.com/aws/aws-lambda-go/lambdacontext"
	"github.com/gkampitakis/go-snaps/snaps"
)

func Test_FullWorkflow_Create_S3_JSON_Simple(t *testing.T) {
	os.Setenv("SOPS_AGE_KEY", "AGE-SECRET-KEY-1EFUWJ0G2XJTJFWTAM2DGMA4VCK3R05W58FSMHZP3MZQ0ZTAQEAFQC6T7T3")
	mocks := &AWS{
		secretsmanager: &SecretsManagerMockClient{
			t: t,
		},
		s3downlaoder: &S3ManagerMockClient{
			t: t,
		},
	}
	d := time.Now().Add(50 * time.Millisecond)
	ctx, _ := context.WithDeadline(context.Background(), d)
	ctx = lambdacontext.NewContext(ctx, &lambdacontext.LambdaContext{
		AwsRequestID:       "AwsRequestID",
		InvokedFunctionArn: "arn:aws:lambda:us-east-2:123456789012:function:cdk-sops-secrets",
	})
	inputJson := ReadJSONFromFile(t, "events/event_create_s3_json_simple.json")
	var event cfn.Event
	err := json.Unmarshal(inputJson, &event)

	phys, data, err := mocks.syncSopsToSecretsmanager(ctx, event)
	check(err)
	snaps.MatchSnapshot(t, ">>>syncSopsToSecretsmanager", phys, data, err)
}

func Test_FullWorkflow_Create_S3_JSON_Complex(t *testing.T) {
	os.Setenv("SOPS_AGE_KEY", "AGE-SECRET-KEY-1EFUWJ0G2XJTJFWTAM2DGMA4VCK3R05W58FSMHZP3MZQ0ZTAQEAFQC6T7T3")
	mocks := &AWS{
		secretsmanager: &SecretsManagerMockClient{
			t: t,
		},
		s3downlaoder: &S3ManagerMockClient{
			t: t,
		},
	}
	d := time.Now().Add(50 * time.Millisecond)
	ctx, _ := context.WithDeadline(context.Background(), d)
	ctx = lambdacontext.NewContext(ctx, &lambdacontext.LambdaContext{
		AwsRequestID:       "AwsRequestID",
		InvokedFunctionArn: "arn:aws:lambda:us-east-2:123456789012:function:cdk-sops-secrets",
	})
	inputJson := ReadJSONFromFile(t, "events/event_create_s3_json_complex.json")
	var event cfn.Event
	err := json.Unmarshal(inputJson, &event)

	phys, data, err := mocks.syncSopsToSecretsmanager(ctx, event)
	check(err)
	snaps.MatchSnapshot(t, ">>>syncSopsToSecretsmanager", phys, data, err)
}

func Test_FullWorkflow_Create_S3_JSON_Complex_StringifyValues(t *testing.T) {
	os.Setenv("SOPS_AGE_KEY", "AGE-SECRET-KEY-1EFUWJ0G2XJTJFWTAM2DGMA4VCK3R05W58FSMHZP3MZQ0ZTAQEAFQC6T7T3")
	mocks := &AWS{
		secretsmanager: &SecretsManagerMockClient{
			t: t,
		},
		s3downlaoder: &S3ManagerMockClient{
			t: t,
		},
	}
	d := time.Now().Add(50 * time.Millisecond)
	ctx, _ := context.WithDeadline(context.Background(), d)
	ctx = lambdacontext.NewContext(ctx, &lambdacontext.LambdaContext{
		AwsRequestID:       "AwsRequestID",
		InvokedFunctionArn: "arn:aws:lambda:us-east-2:123456789012:function:cdk-sops-secrets",
	})
	inputJson := ReadJSONFromFile(t, "events/event_create_s3_json_complex_stringify.json")
	var event cfn.Event
	err := json.Unmarshal(inputJson, &event)

	phys, data, err := mocks.syncSopsToSecretsmanager(ctx, event)
	check(err)
	snaps.MatchSnapshot(t, ">>>syncSopsToSecretsmanager", phys, data, err)
}
func Test_FullWorkflow_Create_S3_JSON_Complex_Flat(t *testing.T) {
	os.Setenv("SOPS_AGE_KEY", "AGE-SECRET-KEY-1EFUWJ0G2XJTJFWTAM2DGMA4VCK3R05W58FSMHZP3MZQ0ZTAQEAFQC6T7T3")
	mocks := &AWS{
		secretsmanager: &SecretsManagerMockClient{
			t: t,
		},
		s3downlaoder: &S3ManagerMockClient{
			t: t,
		},
	}
	d := time.Now().Add(50 * time.Millisecond)
	ctx, _ := context.WithDeadline(context.Background(), d)
	ctx = lambdacontext.NewContext(ctx, &lambdacontext.LambdaContext{
		AwsRequestID:       "AwsRequestID",
		InvokedFunctionArn: "arn:aws:lambda:us-east-2:123456789012:function:cdk-sops-secrets",
	})
	inputJson := ReadJSONFromFile(t, "events/event_create_s3_json_complex_flat.json")
	var event cfn.Event
	err := json.Unmarshal(inputJson, &event)

	phys, data, err := mocks.syncSopsToSecretsmanager(ctx, event)
	check(err)
	snaps.MatchSnapshot(t, ">>>syncSopsToSecretsmanager", phys, data, err)
}
