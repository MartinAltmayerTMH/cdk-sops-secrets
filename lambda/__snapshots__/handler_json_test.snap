
[Test_FullWorkflow_Create_S3_JSON_Simple - 1]
>>>S3MAnagerMockClient.Download.Input
{
  Bucket: "..",
  Key: "../test-secrets/json/sopsfile.enc-age.json"
}
---

[Test_FullWorkflow_Create_S3_JSON_Simple - 2]
>>>SecretsManagerMockClient.PutSecretValue.Input
{
  ClientRequestToken: "../test-secrets/json/sopsfile.enc-age",
  SecretId: "arn:aws:secretsmanager:eu-central-1:123456789012:secret:testsecret",
  SecretString: <sensitive>
}
---

[Test_FullWorkflow_Create_S3_JSON_Simple - 3]
>>>syncSopsToSecretsmanager
arn:custom:sopssync:eu-central-1:123456789012:secret:testsecret
map[string]interface {}{
    "ARN":           "arn:custom:sopssync:eu-central-1:123456789012:secret:testsecret",
    "Name":          "dfb70680d326c245dbdd113d635caa16113e78b1eb8b7447f10eeb861e491724",
    "VersionId":     "46b116fd9409581eb2623320d877cfa1f46cb2487bc3cd0672b1eb95554e65d3",
    "VersionStages": []*string{
        &"dfb70680d326c245dbdd113d635caa16113e78b1eb8b7447f10eeb861e491724",
    },
}
nil
---

[Test_FullWorkflow_Create_S3_JSON_Complex - 1]
>>>S3MAnagerMockClient.Download.Input
{
  Bucket: "..",
  Key: "../test-secrets/json/sopsfile-complex.enc-age.json"
}
---

[Test_FullWorkflow_Create_S3_JSON_Complex - 2]
>>>SecretsManagerMockClient.PutSecretValue.Input
{
  ClientRequestToken: "../test-secrets/json/sopsfile-complex.enc-age",
  SecretId: "arn:aws:secretsmanager:eu-central-1:123456789012:secret:testsecret",
  SecretString: <sensitive>
}
---

[Test_FullWorkflow_Create_S3_JSON_Complex - 3]
>>>syncSopsToSecretsmanager
arn:custom:sopssync:eu-central-1:123456789012:secret:testsecret
map[string]interface {}{
    "ARN":           "arn:custom:sopssync:eu-central-1:123456789012:secret:testsecret",
    "Name":          "dfb70680d326c245dbdd113d635caa16113e78b1eb8b7447f10eeb861e491724",
    "VersionId":     "28e49c30b2d8af77e323b63c5c9e90ec2cf1385ab330eb674695ee065b1252f2",
    "VersionStages": []*string{
        &"dfb70680d326c245dbdd113d635caa16113e78b1eb8b7447f10eeb861e491724",
    },
}
nil
---

[Test_FullWorkflow_Create_S3_JSON_Complex_StringifyValues - 1]
>>>S3MAnagerMockClient.Download.Input
{
  Bucket: "..",
  Key: "../test-secrets/json/sopsfile-complex.enc-age.json"
}
---

[Test_FullWorkflow_Create_S3_JSON_Complex_StringifyValues - 2]
>>>SecretsManagerMockClient.PutSecretValue.Input
{
  ClientRequestToken: "../test-secrets/json/sopsfile-complex.enc-age",
  SecretId: "arn:aws:secretsmanager:eu-central-1:123456789012:secret:testsecret",
  SecretString: <sensitive>
}
---

[Test_FullWorkflow_Create_S3_JSON_Complex_StringifyValues - 3]
>>>syncSopsToSecretsmanager
arn:custom:sopssync:eu-central-1:123456789012:secret:testsecret
map[string]interface {}{
    "ARN":           "arn:custom:sopssync:eu-central-1:123456789012:secret:testsecret",
    "Name":          "dfb70680d326c245dbdd113d635caa16113e78b1eb8b7447f10eeb861e491724",
    "VersionId":     "ddc702937cbed1b68f84659094f8db7c862114642ae2a98baa43c42e16b256f4",
    "VersionStages": []*string{
        &"dfb70680d326c245dbdd113d635caa16113e78b1eb8b7447f10eeb861e491724",
    },
}
nil
---

[Test_FullWorkflow_Create_S3_JSON_Complex_Flat - 1]
>>>S3MAnagerMockClient.Download.Input
{
  Bucket: "..",
  Key: "../test-secrets/json/sopsfile-complex.enc-age.json"
}
---

[Test_FullWorkflow_Create_S3_JSON_Complex_Flat - 2]
>>>SecretsManagerMockClient.PutSecretValue.Input
{
  ClientRequestToken: "../test-secrets/json/sopsfile-complex.enc-age",
  SecretId: "arn:aws:secretsmanager:eu-central-1:123456789012:secret:testsecret",
  SecretString: <sensitive>
}
---

[Test_FullWorkflow_Create_S3_JSON_Complex_Flat - 3]
>>>syncSopsToSecretsmanager
arn:custom:sopssync:eu-central-1:123456789012:secret:testsecret
map[string]interface {}{
    "ARN":           "arn:custom:sopssync:eu-central-1:123456789012:secret:testsecret",
    "Name":          "dfb70680d326c245dbdd113d635caa16113e78b1eb8b7447f10eeb861e491724",
    "VersionId":     "e9a8d3ab5715c1120faae25b669e5905a92fe4b261ae97eb1e65ad9ba52f9ee6",
    "VersionStages": []*string{
        &"dfb70680d326c245dbdd113d635caa16113e78b1eb8b7447f10eeb861e491724",
    },
}
nil
---

[Test_FullWorkflow_Create_INLINE_JSON_Simple - 1]
>>>SecretsManagerMockClient.PutSecretValue.Input
{
  ClientRequestToken: "MyHash",
  SecretId: "arn:aws:secretsmanager:eu-central-1:123456789012:secret:testsecret",
  SecretString: <sensitive>
}
---

[Test_FullWorkflow_Create_INLINE_JSON_Simple - 2]
>>>syncSopsToSecretsmanager
arn:custom:sopssync:eu-central-1:123456789012:secret:testsecret
map[string]interface {}{
    "ARN":           "arn:custom:sopssync:eu-central-1:123456789012:secret:testsecret",
    "Name":          "dfb70680d326c245dbdd113d635caa16113e78b1eb8b7447f10eeb861e491724",
    "VersionId":     "46b116fd9409581eb2623320d877cfa1f46cb2487bc3cd0672b1eb95554e65d3",
    "VersionStages": []*string{
        &"dfb70680d326c245dbdd113d635caa16113e78b1eb8b7447f10eeb861e491724",
    },
}
nil
---

[Test_FullWorkflow_Create_INLINE_JSON_Complex - 1]
>>>SecretsManagerMockClient.PutSecretValue.Input
{
  ClientRequestToken: "MyHash",
  SecretId: "arn:aws:secretsmanager:eu-central-1:123456789012:secret:testsecret",
  SecretString: <sensitive>
}
---

[Test_FullWorkflow_Create_INLINE_JSON_Complex - 2]
>>>syncSopsToSecretsmanager
arn:custom:sopssync:eu-central-1:123456789012:secret:testsecret
map[string]interface {}{
    "ARN":           "arn:custom:sopssync:eu-central-1:123456789012:secret:testsecret",
    "Name":          "dfb70680d326c245dbdd113d635caa16113e78b1eb8b7447f10eeb861e491724",
    "VersionId":     "28e49c30b2d8af77e323b63c5c9e90ec2cf1385ab330eb674695ee065b1252f2",
    "VersionStages": []*string{
        &"dfb70680d326c245dbdd113d635caa16113e78b1eb8b7447f10eeb861e491724",
    },
}
nil
---

[Test_FullWorkflow_Create_INLINE_JSON_Complex_StringifyValues - 1]
>>>SecretsManagerMockClient.PutSecretValue.Input
{
  ClientRequestToken: "MyHash",
  SecretId: "arn:aws:secretsmanager:eu-central-1:123456789012:secret:testsecret",
  SecretString: <sensitive>
}
---

[Test_FullWorkflow_Create_INLINE_JSON_Complex_StringifyValues - 2]
>>>syncSopsToSecretsmanager
arn:custom:sopssync:eu-central-1:123456789012:secret:testsecret
map[string]interface {}{
    "ARN":           "arn:custom:sopssync:eu-central-1:123456789012:secret:testsecret",
    "Name":          "dfb70680d326c245dbdd113d635caa16113e78b1eb8b7447f10eeb861e491724",
    "VersionId":     "ddc702937cbed1b68f84659094f8db7c862114642ae2a98baa43c42e16b256f4",
    "VersionStages": []*string{
        &"dfb70680d326c245dbdd113d635caa16113e78b1eb8b7447f10eeb861e491724",
    },
}
nil
---

[Test_FullWorkflow_Create_INLINE_JSON_Complex_Flat - 1]
>>>SecretsManagerMockClient.PutSecretValue.Input
{
  ClientRequestToken: "MyHash",
  SecretId: "arn:aws:secretsmanager:eu-central-1:123456789012:secret:testsecret",
  SecretString: <sensitive>
}
---

[Test_FullWorkflow_Create_INLINE_JSON_Complex_Flat - 2]
>>>syncSopsToSecretsmanager
arn:custom:sopssync:eu-central-1:123456789012:secret:testsecret
map[string]interface {}{
    "ARN":           "arn:custom:sopssync:eu-central-1:123456789012:secret:testsecret",
    "Name":          "dfb70680d326c245dbdd113d635caa16113e78b1eb8b7447f10eeb861e491724",
    "VersionId":     "e9a8d3ab5715c1120faae25b669e5905a92fe4b261ae97eb1e65ad9ba52f9ee6",
    "VersionStages": []*string{
        &"dfb70680d326c245dbdd113d635caa16113e78b1eb8b7447f10eeb861e491724",
    },
}
nil
---
